src/lang/frontend_ast_validate.rs

```diff
@@ -0,0 +1,803 @@
+//! Validation for the frontend AST
+//!
+//! This module provides semantic validation for parsed AST documents,
+//! including name resolution, arity checking, and wildcard validation.
+
+use std::{collections::HashMap, sync::Arc};
+
+use hex::{FromHex, ToHex};
+
+use crate::{
+    lang::{frontend_ast::*, utils::native_predicate_from_string},
+    middleware::{CustomPredicateBatch, Hash, NativePredicate},
+};
+
+/// A validated AST document with symbol table and diagnostics
+#[derive(Debug, Clone)]
+pub struct ValidatedAST {
+    document: Document,
+    symbols: SymbolTable,
+    diagnostics: Vec<Diagnostic>,
+}
+
+impl ValidatedAST {
+    pub fn document(&self) -> &Document {
+        &self.document
+    }
+
+    pub fn symbols(&self) -> &SymbolTable {
+        &self.symbols
+    }
+
+    pub fn diagnostics(&self) -> &[Diagnostic] {
+        &self.diagnostics
+    }
+
+    pub fn into_document(self) -> Document {
+        self.document
+    }
+}
+
+/// Symbol table containing all predicates and their metadata
+#[derive(Debug, Clone)]
+pub struct SymbolTable {
+    /// All predicates available in this scope
+    pub predicates: HashMap<String, PredicateInfo>,
+    /// Wildcard scopes for each custom predicate
+    pub wildcard_scopes: HashMap<String, WildcardScope>,
+}
+
+/// Information about a predicate
+#[derive(Debug, Clone)]
+pub struct PredicateInfo {
+    pub kind: PredicateKind,
+    pub arity: usize,
+    pub public_arity: usize,
+    pub source_span: Option<Span>,
+}
+
+/// Kind of predicate
+#[derive(Debug, Clone)]
+pub enum PredicateKind {
+    Native(NativePredicate),
+    Custom {
+        index: usize,
+    },
+    BatchImported {
+        batch: Arc<CustomPredicateBatch>,
+        index: usize,
+    },
+    IntroImported {
+        name: String,
+        args_len: usize,
```

Comment:
what's the difference between this value and `PredicateInfo.public_arity`?

Reply:
It is redundant.